src/typed-error.ts

```diff
@@ -0,0 +1,29 @@
+class BaseError {
+    constructor () {
+        Error.apply(this, arguments)
+    }
+}
+BaseError.prototype = Object.create(Error.prototype)
+
+export class TypedError extends BaseError {
+	public name: string
+	public message: string
+	public stack: string
+	constructor(message?: Error | string) {
+		super()
+		let err: Error
+		if (message instanceof Error) {
+			err = message
+		} else {
+			err = new Error(message)
+		}
+		err.name = (this.constructor as any).name
+		this.name = err.name
+		this.message = err.message
+		if (Error.captureStackTrace != null) {
+			Error.captureStackTrace(this, this.constructor)
+		} else if (err.stack != null) {
+			this.stack = err.stack
```

Comment:
It wasn't necessary in one case so I've rejigged the code to avoid creating it in that case as an optimisation, and as an additional optimisation moved the `Error.captureStackTrace` check outside so it can be done just once